config-svc/pkg/api/routes_v1.go

```diff
@@ -15,93 +15,137 @@
-		labels[*data.NameLabel] = data.Name
-	}
-	value := prometheusTargets{
-		{
-			Targets: data.Targets,
-			Labels:  labels,
-		},
+	mgmtPort := 8091
+	if data.CouchbaseConfig.ManagementPort != nil {
+		mgmtPort = int(*data.CouchbaseConfig.ManagementPort)
 	}
-	if value[0].Labels == nil {
-		value[0].Labels = make(map[string]string)
-	}
-	valueJSON, err := json.Marshal(value)
+	cluster, err := couchbase.FetchCouchbaseClusterInfo(
+		data.Hostname,
+		mgmtPort,
+		secure,
+		data.CouchbaseConfig.Username,
+		data.CouchbaseConfig.Password,
+	)
 	if err != nil {
 		return err
 	}
 
-	var path string
-	if s.production {
-		path = prometheusTargetsPathProduction
-	} else {
-		path = prometheusTargetsPathDevelopment
+	staticConfig := prometheus.StaticConfig{
+		Targets: make([]string, len(cluster.Nodes)),
+		Labels: map[string]string{
+			"cluster": cluster.ClusterName,
+		},
 	}
-	path = fmt.Sprintf(path, sanitize.Name(data.Name))
 
-	if data.Overwrite == nil || !*data.Overwrite {
-		_, err := os.Stat(path)
-		if err == nil {
-			return echo.NewHTTPError(http.StatusConflict, "target already exists")
-		} else if errors.Is(err, fs.ErrNotExist) {
-			// Ignore it
-		} else {
+	var anyNodeCB7 bool
+
+	for i, node := range cluster.Nodes {
+		hostname, mgmtPort, err := node.ResolveHostPort(secure)
+		if err != nil {
 			return err
 		}
+		if node.Version.AtLeast(cbvalue.Version7_0_0) {
+			staticConfig.Targets[i] = fmt.Sprintf("%s:%d", hostname, mgmtPort)
+			anyNodeCB7 = true
+		} else {
+			// TODO: allow customising this
+			staticConfig.Targets[i] = fmt.Sprintf("%s:%d", hostname, 9091)
+		}
 	}
 
-	outFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o664)
+	cfgPath := os.Getenv("PROMETHEUS_CONFIG_FILE")
+	if cfgPath == "" {
+		cfgPath = defaultPrometheusConfigPath
+	}
+	cfgFile, err := os.OpenFile(cfgPath, os.O_RDWR, 0)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to open Prometheus config: %w", err)
 	}
-	if _, err = outFile.Seek(0, 0); err != nil {
-		return err
+	defer cfgFile.Close()
+	existingConfig, err := io.ReadAll(cfgFile)
+	if err != nil {
+		return fmt.Errorf("failed to read Prometheus config: %w", err)
 	}
-	if _, err = outFile.Write(valueJSON); err != nil {
-		return err
+	var cfg prometheus.Configuration
+	if err := yaml.Unmarshal(existingConfig, &cfg); err != nil {
+		return fmt.Errorf("failed to parse Prometheus config: %w", err)
+	}
+
+	scrapeConfig := prometheus.ScrapeConfig{
+		// Job name needs to be unique
+		JobName:       fmt.Sprintf("couchbase-server-managed-%d", len(cfg.ScrapeConfigs)+1),
+		StaticConfigs: []prometheus.StaticConfig{staticConfig},
 	}
-	if err = outFile.Close(); err != nil {
+	if anyNodeCB7 {
+		scrapeConfig.HTTPClientConfig = prometheus.HTTPClientConfig{
+			BasicAuth: prometheus.BasicAuthConfig{
+				Username: data.CouchbaseConfig.Username,
+				Password: data.CouchbaseConfig.Password,
+			},
+		}
+	}
+
+	cfg.ScrapeConfigs = append(cfg.ScrapeConfigs, &scrapeConfig)
```

Comment:
I think this is a bit too much coupling between essentially file handling and in-memory scrape config. It would be better to split it out slightly and add some more unit tests - I don't think you need to worry as much about the file handling but certainly the various scrape configs as an in-memory array would be useful to test different combinations are correctly created.

Reply:
To a certain extent, it is already split out - `prometheus.Configuration` is a plain old Go structure with special YAML marshal/unmarshal behaviour, it leaves handling the YAML output to its caller. Or do you mean splitting up the API handler (`routes_v1.go`)?

Reply:
Yeah sorry more the flow of this method itself, it smells a bit to me.